Feature/bluetooth detection

carlsve opened the pull request:
note:
- Error handling from Bluez to DetectionError is done poorly, needs fix for more descriptive errors.
- RSSI distance is provided, can we find a way to make RSSI more meaningful? (Provide RSSI to distance calculation and send along with the discord app)

- I am unsure about how to deal with the cargo.lock and cargo.toml, don't merge without fixing this first
- The example code used block_on from async_std, while our project uses tokio. Two async dependencies are too much, how can we implement block_on in tokio? (less important for prototype, but still discussible)

carlsve committed:
Change Detector::wait_for_device to async

carlsve committed:
Enable bluetooth device detection

ysndr commented on src/bluetooth/detector.rs:
```diff
@@ -0,0 +1,118 @@
+                    rssi,
+                    ..
+                } => {
+                    let device = SimpleDevice(address.to_string());
```
Having to construct the device already seems a bit unfortunate. I was planning to get the actual device from the registry. This way we can add meta data to the device such as distance conversion parameters and friendly names.
Not much of an issue here, yet but that should be fixed eventually.

ysndr replied:
An `address` can be passed to registry directly to check whether a device is associated with it

ysndr commented on src/bluetooth/detector.rs:
```diff
@@ -0,0 +1,118 @@
+        let (controller, info) = controllers
+            .into_iter()
+            .filter_map(|controller| {
+                let info = block_on(client.get_controller_info(controller)).ok()?;
```
https://docs.rs/tokio/1.6.0/tokio/task/fn.block_in_place.html

This looks like how it's done with tokio

ysndr commented on src/bluetooth/detector.rs:
```diff
@@ -0,0 +1,118 @@
+                }
+            })
+            .nth(0)
+            .expect("no usable controllers found");
```
An error Type would be make sense here, just to not have it panic..

ysndr commented on src/bluetooth/detector.rs:
```diff
@@ -0,0 +1,118 @@
+        
+        println!("found controller {}", controller);
+
+        if !info.current_settings.contains(ControllerSetting::Powered) {
```
Does the discovery itself require sudo as well or only the power up part?

ysndr commented on src/bluetooth/detector.rs:
```diff
@@ -0,0 +1,118 @@
+            .await?;
+
+        // just wait for discovery forever
+        loop {
```
Okay thai works on the same thread. I was worried one needs to start it in a new thread and wait for some event handler to return with data.

ysndr commented on Cargo.toml:
```diff
@@ -13,15 +13,17 @@ anyhow="1.0"
 cfg-if="1.0"
 futures="0.3"
-
+bluez = { version = "0.3.1", optional = true }
```
I'd stick to minor releases so "0.3" should be enough, cargo will figure out which version it wants

ysndr commented on src/bluetooth/detector.rs:
```diff
@@ -0,0 +1,118 @@
+            .nth(0)
+            .expect("no usable controllers found");
+        
+        println!("found controller {}", controller);
```
It's fine to use the log crate,
So print  this data as info!(...)

ysndr commented on src/bluetooth/detector.rs:
```diff
@@ -0,0 +1,118 @@
+    fn from(err: BluezError) -> Self {
+        DetectorError::EOLError
+    }
+}
```
```suggestion
pub enum DetectorError {
    #[error("Error in bluez: {0}")]
    Bluez(#[from] BluezError),

    #[error("Couldn't find supported Bluetooth device")]
    NoDevice
}

impl ServiceError for DetectorError {}
```

ysndr reviewed (changes requested):
Looks fine, I am not surprised that it works already. Good job!
My concerns are the same as yours.
- Error type
- async_std
And some minor details.
I left some comments on the code as pointers.

ysndr commented:
RSSI -> Distance would require device specific metadata.
Finally we should be able to configure a minimum distance, possibly also exposing it to cli/env

ysndr committed:
Apply some suggested changes (1)

println -> log::_

ysndr committed:
Apply some suggested changes (2)

Error Type

ysndr committed:
Apply some suggested changes (3)

Remove dependency on async_std

ysndr committed:
General Formatting

ysndr committed:
Specify bluez dependency by minor version only

ysndr commented on src/bluetooth/detector.rs:
```diff
@@ -0,0 +1,127 @@
+            warn!("Bluetooth controller {} powered off", controller);
+            info!("Powering on bluetooth controller {}", controller);
+            client.set_powered(controller, true).await?;
+        }
```
I kinda want to put the device finding part into `new()` and the boot up in its own method too, combined with typestates this could provide a safer interface to starting the Bluetooth service

ysndr committed:
Add proper new line ending

carlsve committed:
Change filter_map to naive for loop to skip runtime spawning issues.

carlsve committed:
Remove unused imports in bluetooth/detector

carlsve committed:
Small fix in main to skip having to return result from register_device.

ysndr committed:
Pass Actuator as refernce

ysndr committed:
Nix flakify

ysndr committed:
Abstract device

- Registry as HashMap
- Remove Trait bounds to support any Device in authenticators
- Implement Bluetooth device which can hold metadata

ysndr commented:
PR now also implements flake support and abstracts the registry to a hashmap of arbitrary device types

ysndr commented:
Cargo check is happy
Clippy not so much

ysndr committed:
Refactor client creation to `new`

ysndr commented on src/bluetooth/mod.rs:
```diff
@@ -0,0 +1 @@
+pub mod detector;
```
@karlpet can you please enable adding newlines in your editor?
I don't think it will be an issue for us but GitHub complains :)
